sparsity/dask/core.py

```diff
@@ -518,4 +528,179 @@ def apply_and_enforce(func, arg, kwargs, meta):
+    if len(npartitions) > 1:
+        raise ValueError("All arguments must have same number of partitions")
+    npartitions = npartitions.pop()
+
+    if split_every is None:
+        split_every = 8
+    elif split_every is False:
+        split_every = npartitions
+    elif split_every < 2 or not isinstance(split_every, int):
+        raise ValueError("split_every must be an integer >= 2")
+
+    token_key = tokenize(token or (chunk, aggregate), meta, args,
+                         chunk_kwargs, aggregate_kwargs, combine_kwargs,
+                         split_every, split_out, split_out_setup,
+                         split_out_setup_kwargs)
+
+    # Chunk
+    a = '{0}-chunk-{1}'.format(token or funcname(chunk), token_key)
+    if len(args) == 1 and isinstance(args[0], SparseFrame) and not chunk_kwargs:
+        dsk = {(a, 0, i, 0): (chunk, key)
+               for i, key in enumerate(args[0].__dask_keys__())}
+    else:
+        dsk = {(a, 0, i, 0): (apply, chunk,
+                              [(x._name, i) if isinstance(x, SparseFrame)
+                               else x for x in args], chunk_kwargs)
+               for i in range(args[0].npartitions)}
+
+    # Split
+    # this splits the blocks (usually) by their index and
+    # basically performs a task sort such that the next tree
+    # aggregation will result in the disered number of partitions
+    # given by the split_out parameter
+    if split_out and split_out > 1:
+        split_prefix = 'split-%s' % token_key
+        shard_prefix = 'shard-%s' % token_key
+        for i in range(args[0].npartitions):
+            # For now we assime that split_out_setup selects the index
+            # as we will only support index groupbys for now. So we can
+            # use the function provided by dask.
+            dsk[(split_prefix, i)] = (hash_shard, (a, 0, i, 0), split_out,
+                                      split_out_setup, split_out_setup_kwargs)
+            # at this point we have dictionaries of dataframes the key
+            # is the to which output partition this dataframe belongs to
+            # the next line unpacks this dictionaries into pure dataframes again
+            # now with the correct dask key for their partition
+            for j in range(split_out):
+                dsk[(shard_prefix, 0, i, j)] = (getitem, (split_prefix, i), j)
+        a = shard_prefix
+    else:
+        split_out = 1
+
+    # Combine
+    b = '{0}-combine-{1}'.format(token or funcname(combine), token_key)
+    k = npartitions
+    depth = 0
+    while k > split_every:
+        for part_i, inds in enumerate(partition_all(split_every, range(k))):
+            for j in range(split_out):
+                # concat basically just concatenates partitions along
+                # the first dimension. Will replace this with the vstack
+                # method from the SparseFrame class. Dask does some
+                # extra checks which we can hopefully neglect.
```

Comment:
I think this comment is not necessary anymore as it's already using `vstack`.
Dask checks can be in fact neglected? (I have no idea which checks are these)

Reply:
Ok removed the comment. Dask does some extra checks at this point e.g. if the columns are the same but vstack does that too.